src/raytracing.jl

```diff
@@ -70,18 +71,74 @@ normal_to_barrel(c::Cylinder, p:: Vector{Float64}) = [p[1], p[2], 0.0] ./c.r
 
 equation of cylynder: F(x,y,z) = x^2 + y^2 - r^2 = 0
 """
-cylinder_equation(c::Cylinder, p:: Vector{Float64}) = p[1]^2 + p[2]^2 - c.r^2
+cylinder_equation(c::Cylinder, p::Vector{Float64}) = p[1]^2 + p[2]^2 - c.r^2
 
 
-length(c::Cylinder) = abs.(c.zmax - c.zmin)
+cylinder_length(c::Cylinder) = abs.(c.zmax - c.zmin)
     
 perimeter(c::Cylinder) = 2 * π * c.r
 
-area_barrel(c::Cylinder) = 2 * π * c.r * length(c)
+area_barrel(c::Cylinder) = 2 * π * c.r * cylinder_length(c)
     
 area_endcap(c::Cylinder) = π * c.r^2
     
 area(c::Cylinder) = area_barrel(c) + 2 * area_endcap(c)
 
-volume(c::Cylinder)= area_endcap(c) * length(c)
+volume(c::Cylinder)= area_endcap(c) * cylinder_length(c)
+
+
+"""Returns True if point in end-caps of cyinder"""
+in_endcaps(c:: Cylinder, p::Vector{Float64}) = isapprox(p[3], c.zmin) || isapprox(p[3], c.zmax)
+    
+    
+"""Computes intersection between a ray and a cylinder"""
+function propagate(r::Ray, cy::Cylinder)
+    
+    # intersection roots 
+    a = r.d[1]^2 + r.d[2]^2
+    b = 2 * (r.e[1] * r.d[1] + r.e[2] * r.d[2])
+    c = r.e[1]^2 + r.e[2]^2 - cy.r^2
+
+    # xray(t) propagates ray along t 
+    xray = ATools.propagate_ray(r)
+
+    # If no real roots return nothing 
+    arg = b^2 - 4*a*c
+
+    # two possible roots 
+    t1 = (-b + sqrt(arg))/(2*a)
+    t2 = (-b - sqrt(arg))/(2*a)
+
+    # Z position of the two roots 
+    z1 = r.e[3] + t1 * r.d[3]
+    z2 = r.e[3] + t2 * r.d[3]
+
+    # belong-to-cylinder condition 
+    zc1 = cy.zmin < z1 < cy.zmax
+    zc2 = cy.zmin < z2 < cy.zmax
```

Comment:
It's not clear to me exactly what the function is meant to do. You keep both `t1` and `t2` which made me think that you wanted intersections for back-to-back gammas but then the function only returns one position.

If you assume that the origin is inside the cylinder (should be true in our application), positive t is the only relevant intersection if we only want to follow the ray. If both are possible but only one returned the meaning of the function is ambiguous. If you only want the forward intersections then a lot of the `if else` below can be removed in favour of taking the positive t. If you want a function that gives you the two/one/or no intersection(s) of the back-to-back gammas then I'd return both with `nothing` or `missing` in the case of no valid intersection.

Reply:
https://www.cl.cam.ac.uk/teaching/1999/AGraphHCI/SMAG/node2.html

Reply:
I understand the calculations for where the intersections should be. My question is what you want to return. If you only want the forward going ray then you can check if one of the `t` is positive, if you only want to consider origins within the cylinder.

In the file you link they already talk about `t > 0` being the only really relevant root for a ray my question about what you want is inspired by the back-to-back nature of the specific problem the code is written for.

Reply:
Let me see if I follow you. The argument is that the solution with t<0 also works, given that there are two back-to-back gammas?

Reply:
If you assume back-to-back gammas both `t > 0` and `t < 0` can be valid (if you are inside the cylinder you can only have one positive and 1 negative or neither defined). In the case that you want both I'd check each intersect for validity (intersect within the limits of interest) and return them both with something like a `missing` or `nothing` for any of them that aren't valid. If you're only interested in the forward ray (`t > 0`) you can already ignore one of them from earlier in the function and save yourself some of the `if else`